db/map.go

```diff
@@ -0,0 +1,319 @@
+	return WrappedExecutor{SqlExecutor: m.DbMap}.Delete(list...)
+}
+
+func (m *WrappedMap) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	return WrappedExecutor{SqlExecutor: m.DbMap}.Select(holder, query, args...)
+}
+
+func (m *WrappedMap) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	return WrappedExecutor{SqlExecutor: m.DbMap}.SelectOne(holder, query, args...)
+}
+
+func (m *WrappedMap) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return WrappedExecutor{SqlExecutor: m.DbMap}.Exec(query, args...)
+}
+
+func (m *WrappedMap) WithContext(ctx context.Context) gorp.SqlExecutor {
+	return WrappedExecutor{SqlExecutor: m.DbMap.WithContext(ctx)}
+}
+
+func (m *WrappedMap) Begin() (Transaction, error) {
+	tx, err := m.DbMap.Begin()
+	if err != nil {
+		return tx, ErrDatabaseOp{
+			Op:  "begin transaction",
+			Err: err,
+		}
+	}
+	return WrappedTransaction{
+		Transaction: tx,
+	}, err
+}
+
+// WrappedTransaction wraps a *gorp.Transaction such that its major functions
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedTransaction struct {
+	*gorp.Transaction
+}
+
+func (tx WrappedTransaction) WithContext(ctx context.Context) gorp.SqlExecutor {
+	return WrappedExecutor{SqlExecutor: tx.Transaction.WithContext(ctx)}
+}
+
+func (tx WrappedTransaction) Commit() error {
+	return tx.Transaction.Commit()
+}
+
+func (tx WrappedTransaction) Rollback() error {
+	return tx.Transaction.Rollback()
+}
+
+func (tx WrappedTransaction) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Get(holder, keys...)
+}
+
+func (tx WrappedTransaction) Insert(list ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Insert(list...)
+}
+
+func (tx WrappedTransaction) Update(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Update(list...)
+}
+
+func (tx WrappedTransaction) Delete(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Delete(list...)
+}
+
+func (tx WrappedTransaction) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Select(holder, query, args...)
+}
+
+func (tx WrappedTransaction) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).SelectOne(holder, query, args...)
+}
+
+func (tx WrappedTransaction) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Exec(query, args...)
+}
+
+var (
+	// selectTableRegexp matches the table name from an SQL select statement
+	selectTableRegexp = regexp.MustCompile(`(?i)^\s*select\s+[a-z\d\.\(\), \_\*` + "`" + `]+\s+from\s+([a-z\d\_,` + "`" + `]+)`)
```

Comment:
Since these regexes need to use ``` ` ``` and don't need to use `"`, it'll probably be clearer to quote them with `"` and drop the concatenations.

Reply:
Won't that require double escaping lots of other things (e.g. `\s`?)

I was not thrilled to realize there's no way to put a backtick literal into a Golang raw string but using a regular string seemed like it caused more problems than it solved.

Reply:
Oh right, good point. I had missed that part.